spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraRDDPartitionGenerator.scala

```diff
@@ -0,0 +1,323 @@
+    val replicas = metadata.getReplicas(Metadata.quote(keyspaceName), range).map(_.getAddress).toSet
+    val dataSize = (tokenFactory.ringFraction(startToken, endToken) * totalDataSize).toLong
+    new TokenRange(startToken, endToken, replicas, dataSize)
+  }
+
+  def fullRange: TokenRange = {
+
+    val replicas = connector
+      .withClusterDo{cluster => cluster.getMetadata}
+      .getAllHosts
+      .map(_.getAddress)
+      .toSet
+
+    new TokenRange(
+      tokenFactory.minToken,
+      tokenFactory.maxToken,
+      replicas,
+      0)
+  }
+
+  private def describeRing: Seq[TokenRange] = {
+    connector.withClusterDo { cluster =>
+      val metadata = cluster.getMetadata
+      for (tr <- metadata.getTokenRanges.toSeq) yield tokenRange(tr, metadata)
+    }
+  }
+
+  private def splitsOf(
+      tokenRanges: Iterable[TokenRange],
+      splitter: TokenRangeSplitter[V, T]): Iterable[TokenRange] = {
+
+    val parTokenRanges = tokenRanges.par
+    parTokenRanges.tasksupport = new ForkJoinTaskSupport(CassandraRDDPartitionGenerator.pool)
+    (for (tokenRange <- parTokenRanges;
+          split <- splitter.split(tokenRange, splitSize)) yield split).seq
+  }
+
+  private def splitToCqlClause(range: TokenRange): Iterable[CqlTokenRange] = {
+    val startToken = range.start.value
+    val endToken = range.end.value
+    val wrapAround = range.isWrapAround
+    val pk = tableDef.partitionKey.map(_.columnName).map(quote).mkString(", ")
+
+    if (range.end == tokenFactory.minToken)
+      List(CqlTokenRange(GreaterThan(pk), startToken))
+    else if (range.start == tokenFactory.minToken)
+      List(CqlTokenRange(LessThanOrEquals(pk), endToken))
+    else if (!range.isWrapAround)
+      List(CqlTokenRange(GreaterThanAndLessThanOrEquals(pk), startToken, endToken))
+    else
+      List(
+        CqlTokenRange(GreaterThan(pk), startToken),
+        CqlTokenRange(LessThanOrEquals(pk), endToken))
+  }
+
+  private def createTokenRangeSplitter: TokenRangeSplitter[V, T] = {
+    tokenFactory.asInstanceOf[TokenFactory[_, _]] match {
+      case TokenFactory.RandomPartitionerTokenFactory =>
+        new RandomPartitionerTokenRangeSplitter(totalDataSize).asInstanceOf[TokenRangeSplitter[V, T]]
+      case TokenFactory.Murmur3TokenFactory =>
+        new Murmur3PartitionerTokenRangeSplitter(totalDataSize).asInstanceOf[TokenRangeSplitter[V, T]]
+      case _ =>
+        throw new UnsupportedOperationException(s"Unsupported TokenFactory $tokenFactory")
+    }
+  }
+
+  lazy val partitions = partitionsAndRanges.map(_._1)
+  lazy val ranges = partitionsAndRanges.map(x =>
+    (x._2, x._3.map( range =>
+      (range.start, range.end, range.isWrapAround))))
+
+
+  /**
+    * Generates a list of the CassandraPartitions to be used in the TableScanRDD and a list
+    * of the TokenRanges used to craft those partitions along with their indexes.
+    */
+  lazy val partitionsAndRanges: Seq[(CassandraPartition, Int, Seq[TokenRange])] = {
```

Comment:
This should most likely be refactored, The Index is not really needed (i'm just paranoid about the Ordering. The real issue here is that going from TokenRanges to CassandraPartitions also requires RowCount and Replicas. This means we can't just go from ranges to CassandraPartitions, The Cassandra partitions lose all of their type and have varargs which makes them more annoying to convert back to Ranges. 

Maybe the right thing to do is to make a super class PartitionRangeInformation[V](Seq[(V, V , Boolean)], RowCount, Replicas) ? This could be converted directly to a CassandraPartition